### PR TITLE
Wandb logging

### PR DESCRIPTION
**Pull request by maxlampe**
# Pull Request

## Description

I've added wandb logging and ported all old logging metrics to new codebase.

I made a two normative changes that should be discussed/red-teamed:
- step functions (i.e., Algorithm.step() and Harness.step()) now return the loss and [new] a dict of log entries.
- values to log are set when defining the step function for each algorithm (e.g., dpo, ipo, ...) and can also be set in the application in the step for loop (ast_hf.py, ast_basic.py) to add global parameters (e.g., step number) or custom user things without going into the weeds.

I've marked everything to check or that is unresolved with a "#TODO: [...]" in the code.

Hope you are having a good day <3

### CODE REVIEW

File: examples/ast_basic.py [hunk 1]
```diff
@@ -120,32 +123,46 @@ def __get_logprobs(self, model, context, continuation):
...
+    env = ASTEnvironment(problem, PROMPTS)
+
+    # instantiate our solution
+    solver = DPO(problem)
+    optimizer = AdamW(problem.parameters(), lr=1e-5)
+
+    # this is a training harness, from which we can call various functions to
+    # handle training details
+    harness = Harness(
+        env, solver, batch_size=4, num_episodes_per_experience=2, use_wandb=True
```
[the remark this answers is not in the document]
Comment (maxlampe) [reply examples/ast_basic.py:145]: That would be better for flexibility. Will make it so.

File: src/astra_rl/training/harness.py [hunk 2]
```diff
@@ -1,8 +1,12 @@
 from typing import Generic, Sequence, Optional, Dict, Any, Iterator
+from abc import abstractmethod
+import os
 
 import torch
 from torch.utils.data import Dataset, DataLoader
 
+import wandb
```
[the remark this answers is not in the document]
Comment (maxlampe) [reply src/astra_rl/training/harness.py:8]: I am generally not a fan of inline imports for style/good practice reasons (keep imports always at the top for readability and transparency). How about a try: import at the top? Does this mean we should also remove it form the project requirements/uv.lock file?

File: src/astra_rl/training/harness.py [hunk 3]
```diff
@@ -87,6 +91,8 @@ def __init__(
         environment: Environment[StateT, ActionT],
         algorithm: Algorithm[StateT, ActionT, Step, Batch],
         num_episodes_per_experience: int = 32,
+        use_wandb: bool = True,
+        wandb_kwargs: Optional[Dict[str, Any]] = None,
         **kwargs: Any,
```
[the remark this answers is not in the document]
Comment (maxlampe) [reply src/astra_rl/training/harness.py:96]: great idea and makes sense! that would be clearer :)

File: src/astra_rl/training/harness.py [hunk 4]
```diff
@@ -148,3 +168,34 @@ def experience(self, seed: Optional[int] = None) -> Iterator[Batch]:
...
+            if "WANDB_API_KEY" not in os.environ:
+                raise EnvironmentError(
+                    "WANDB_API_KEY environment variable is not set. Please set it to use Weights & Biases."
+                )
+            # TODO: This was a hotfix, but all other ways failed me (Max)
+            run = wandb.init(project="astra_rl", config=self.wandb_kwargs)  # type: ignore[attr-defined]
+            return run
+        return None
+
+    @abstractmethod
```
[the remark this answers is not in the document]
Comment (maxlampe) [reply src/astra_rl/training/harness.py:189]: This was an error as I initially had this not using class attributes. Will fix now.

File: src/astra_rl/algorithms/dpo.py [hunk 5]
```diff
@@ -133,4 +162,34 @@ def step(self, batch: DPOBatch[StateT, ActionT]) -> torch.Tensor:
...
 
-        return loss.mean()
+        # Calculate addition quantities
+        # TODO: CHECK ME for correctness and completion!
+        chosen_rewards = self.beta * (attacker_logprobs_win - baseline_logprobs_win)
+        rejected_rewards = self.beta * (attacker_logprobs_loss - baseline_logprobs_loss)
+        reward_accuracies = (chosen_rewards > rejected_rewards).float()
+        reward_margin = chosen_rewards - rejected_rewards
+
+        logging_dict: dict[Any, Any] = {
```
[the remark this answers is not in the document]
Comment (Jemoka) [reply src/astra_rl/algorithms/dpo.py:172]: Entirely fair; how's `Dict[str, Any]`? We can then (if we want, but totally optional!) have runtime asserts for things that can be `wandb` logged (i.e., numbers, or `wandb.Html`, etc.)

**Comment by maxlampe** [comment]
@Jemoka (Just to recap what we've talked about): 

- Separate wandb logging from Harness class into separate class
- call new logger class in Harness
- add docstring/comment line in application script ast_basic to instruct people to use their own logger there
- [when trainer class ready] add new loggign class to trainer object

File: src/astra_rl/training/harness.py [hunk 6]
```diff
@@ -100,9 +106,14 @@ def __init__(
         self.environment = environment
         self.algorithm = algorithm
         self.num_episodes_per_experience = num_episodes_per_experience
+        self.use_wandb = use_wandb
+        self.wandb_kwargs = wandb_kwargs or {}
```
[the remark this answers is not in the document]
Comment (Jemoka) [reply src/astra_rl/training/harness.py:107]: sounds good

**Review by Jemoka — APPROVED** [review]
LGTM! Feel free to mash that merge button 🚢
